<a href="https://colab.research.google.com/github/Atul57/Data-Science-project/blob/main/Assignment_4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import pandas as pd
import numpy as np
from google.colab import drive
drive.mount('/content/drive')
'/content/drive/MyDrive/student_data.csv'

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


'/content/drive/MyDrive/student_data.csv'

Q.1 = Create the DataFrame from
the table above. Display:
shape, columns, dtypes,
head(5), and tail(3).

In [21]:
df = pd.read_csv('/content/drive/MyDrive/student_data.csv')
# Shape
df.shape

(10, 8)

In [22]:
df.columns

Index(['StudentID', 'Name', 'Age', 'City', 'Department', 'Score', 'Grade',
       'Passed'],
      dtype='object')

In [20]:
df.dtypes

,0
StudentID,object
Name,object
Age,int64
City,object
Department,object
Score,int64
Grade,object
Passed,bool


In [19]:
df.head(5)

,StudentID,Name,Age,City,Department,Score,Grade,Passed
0,S001,Arjun Mehta,21,Delhi,CSE,78,B,True
1,S002,Divya Sharma,22,Mumbai,ECE,91,A,True
2,S003,Karan Patel,20,Pune,CSE,55,C,True
3,S004,Meera Nair,23,Chennai,MECH,42,D,True
4,S005,Sahil Gupta,21,Delhi,ECE,88,A,True


In [18]:
df.tail(3)

,StudentID,Name,Age,City,Department,Score,Grade,Passed
7,S008,Pooja Yadav,23,Chennai,CSE,92,A,True
8,S009,Vivek Kumar,21,Pune,ECE,71,B,True
9,S010,Nisha Agarwal,22,Delhi,MECH,58,C,True


Q. 2 = Run df.info() and
df.describe(). From
describe(): what is the
mean Score? What is the
standard deviation?

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10 entries, 0 to 9
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   StudentID   10 non-null     object
 1   Name        10 non-null     object
 2   Age         10 non-null     int64 
 3   City        10 non-null     object
 4   Department  10 non-null     object
 5   Score       10 non-null     int64 
 6   Grade       10 non-null     object
 7   Passed      10 non-null     bool  
dtypes: bool(1), int64(2), object(5)
memory usage: 702.0+ bytes


In [24]:
df.describe()

,Age,Score
count,10.000000,10.000000
mean,21.500000,67.700000
std,1.080123,20.133168
min,20.000000,35.000000
25%,21.000000,55.750000
50%,21.500000,69.000000
75%,22.000000,85.500000
max,23.000000,92.000000


Mean Score = 67.700000
Standard Deviation = 20.133168

Q. 3 = Access the 'Name' column
using df['Name']. Then
access 'Score' and 'Grade'
columns together using
double brackets.

In [27]:
df['Name']

,Name
0,Arjun Mehta
1,Divya Sharma
2,Karan Patel
3,Meera Nair
4,Sahil Gupta
5,Anjali Singh
6,Rohan Verma
7,Pooja Yadav
8,Vivek Kumar
9,Nisha Agarwal


In [29]:
df[['Score', 'Grade']]

,Score,Grade
0,78,B
1,91,A
2,55,C
3,42,D
4,88,A
5,35,F
6,67,B
7,92,A
8,71,B
9,58,C


Q . 4 = Use df.loc to select: (a) row
at index 0, (b) rows 2 to 5,
(c) rows 0–4 but only
columns Name, Score,
Grade

In [31]:
df.loc[0]

,0
StudentID,S001
Name,Arjun Mehta
Age,21
City,Delhi
Department,CSE
Score,78
Grade,B
Passed,True


In [32]:
df.loc[2:5]

,StudentID,Name,Age,City,Department,Score,Grade,Passed
2,S003,Karan Patel,20,Pune,CSE,55,C,True
3,S004,Meera Nair,23,Chennai,MECH,42,D,True
4,S005,Sahil Gupta,21,Delhi,ECE,88,A,True
5,S006,Anjali Singh,22,Jaipur,CSE,35,F,False


In [36]:
df.loc[0:4, ['Name', 'Score', 'Grade']]

,Name,Score,Grade
0,Arjun Mehta,78,B
1,Divya Sharma,91,A
2,Karan Patel,55,C
3,Meera Nair,42,D
4,Sahil Gupta,88,A


Q.5 = Use df.iloc to select: (a)
first 3 rows, (b) last 2 rows,
(c) rows at positions 1, 3, 5
and columns at positions 0,
2, 4.

In [37]:
df.iloc[0:3]

,StudentID,Name,Age,City,Department,Score,Grade,Passed
0,S001,Arjun Mehta,21,Delhi,CSE,78,B,True
1,S002,Divya Sharma,22,Mumbai,ECE,91,A,True
2,S003,Karan Patel,20,Pune,CSE,55,C,True


In [43]:
df.iloc[-2:]

,StudentID,Name,Age,City,Department,Score,Grade,Passed
8,S009,Vivek Kumar,21,Pune,ECE,71,B,True
9,S010,Nisha Agarwal,22,Delhi,MECH,58,C,True


In [44]:
df.iloc[[1, 3, 5], [0, 2, 4]]

,StudentID,Age,Department
1,S002,22,ECE
3,S004,23,MECH
5,S006,22,CSE


Q . 6 = Filter students where Score > 80. How many students scored above 80?

In [45]:
students_above_80 = df[df['Score'] > 80]
display(students_above_80)
num_students_above_80 = len(students_above_80)
print(f"Number of students who scored above 80: {num_students_above_80}")

,StudentID,Name,Age,City,Department,Score,Grade,Passed
1,S002,Divya Sharma,22,Mumbai,ECE,91,A,True
4,S005,Sahil Gupta,21,Delhi,ECE,88,A,True
7,S008,Pooja Yadav,23,Chennai,CSE,92,A,True


Number of students who scored above 80: 3


Q . 7 = Filter students from Delhi
who are in CSE
department. Show Name,
Score, and Grade only.


In [46]:
filtered_students = df[(df['City'] == 'Delhi') & (df['Department'] == 'CSE')]
display(filtered_students[['Name', 'Score', 'Grade']])

,Name,Score,Grade
0,Arjun Mehta,78,B


Q . 8 = Find all students who
Failed (Passed == False).
What is their average
Score?

In [47]:
failed_students = df[df['Passed'] == False]
display(failed_students)

average_score_failed = failed_students['Score'].mean()
print(f"Average score of students who failed: {average_score_failed:.2f}")

,StudentID,Name,Age,City,Department,Score,Grade,Passed
5,S006,Anjali Singh,22,Jaipur,CSE,35,F,False


Average score of students who failed: 35.00


Q . 9 = Filter students whose
Name contains the string
'a' (case-insensitive) using
.str.contains().

In [48]:
students_with_a_in_name = df[df['Name'].str.contains('a', case=False)]
display(students_with_a_in_name)

,StudentID,Name,Age,City,Department,Score,Grade,Passed
0,S001,Arjun Mehta,21,Delhi,CSE,78,B,True
1,S002,Divya Sharma,22,Mumbai,ECE,91,A,True
2,S003,Karan Patel,20,Pune,CSE,55,C,True
3,S004,Meera Nair,23,Chennai,MECH,42,D,True
4,S005,Sahil Gupta,21,Delhi,ECE,88,A,True
5,S006,Anjali Singh,22,Jaipur,CSE,35,F,False
6,S007,Rohan Verma,20,Mumbai,MECH,67,B,True
7,S008,Pooja Yadav,23,Chennai,CSE,92,A,True
8,S009,Vivek Kumar,21,Pune,ECE,71,B,True
9,S010,Nisha Agarwal,22,Delhi,MECH,58,C,True


Q . 10 = Add a new column
'Score_Category': if Score
>= 80 → 'Distinction', Score
>= 60 → 'Merit', else
'Pass/Fail'. Use np.where()
or nested condition.


In [49]:
df['Score_Category'] = np.where(df['Score'] >= 80, 'Distinction',
                                  np.where(df['Score'] >= 60, 'Merit', 'Pass/Fail'))
display(df)

,StudentID,Name,Age,City,Department,Score,Grade,Passed,Score_Category
0,S001,Arjun Mehta,21,Delhi,CSE,78,B,True,Merit
1,S002,Divya Sharma,22,Mumbai,ECE,91,A,True,Distinction
2,S003,Karan Patel,20,Pune,CSE,55,C,True,Pass/Fail
3,S004,Meera Nair,23,Chennai,MECH,42,D,True,Pass/Fail
4,S005,Sahil Gupta,21,Delhi,ECE,88,A,True,Distinction
5,S006,Anjali Singh,22,Jaipur,CSE,35,F,False,Pass/Fail
6,S007,Rohan Verma,20,Mumbai,MECH,67,B,True,Merit
7,S008,Pooja Yadav,23,Chennai,CSE,92,A,True,Distinction
8,S009,Vivek Kumar,21,Pune,ECE,71,B,True,Merit
9,S010,Nisha Agarwal,22,Delhi,MECH,58,C,True,Pass/Fail


Sort the DataFrame by
Score in descending order.
Reset the index. Show the
top 5 students.

In [50]:
df_sorted = df.sort_values(by='Score', ascending=False).reset_index(drop=True)
display(df_sorted.head(5))

,StudentID,Name,Age,City,Department,Score,Grade,Passed,Score_Category
0,S008,Pooja Yadav,23,Chennai,CSE,92,A,True,Distinction
1,S002,Divya Sharma,22,Mumbai,ECE,91,A,True,Distinction
2,S005,Sahil Gupta,21,Delhi,ECE,88,A,True,Distinction
3,S001,Arjun Mehta,21,Delhi,CSE,78,B,True,Merit
4,S009,Vivek Kumar,21,Pune,ECE,71,B,True,Merit
